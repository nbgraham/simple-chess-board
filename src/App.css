#root {
  position: absolute;
  height: 100%;
  width: 100%;
}

.flex-grid {
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  height: 100%;
  width: 100%;
}
.flex-grid .col {
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}
.flex-grid .header {
  flex: .2;
  display: flex;
  align-items: center;
  justify-content: center;
}
.flex-grid .cell {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
}

.cell.boardColorblack {
  background-color: #c97725;
  border: 3px #c97725 solid;
  color: white;
}

.cell.boardColorwhite {
  background-color: #783c00;
  border: 3px #783c00 solid;
  color: white;
}

.cell.selected {
  border: 3px red solid;
}

.cell.available {
  border: 3px green dotted;
}

.cell.lastMoved {
  border: 3px blue dashed;
}

.App {
  text-align: center;
}