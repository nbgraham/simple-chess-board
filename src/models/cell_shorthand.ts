import { Cell } from "./cell";
import { BoardUtils } from "../utils/board_utils";

export const c = (cellName: string): Cell => ({ 
    rowIndex: BoardUtils.N_ROWS - Number(cellName[1]), 
    columnIndex: BoardUtils.LETTERS_FOR_COLUMNS.indexOf(cellName[0])
})