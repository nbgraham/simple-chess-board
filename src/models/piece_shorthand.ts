import { TPiece, PieceType } from './piece';
import { BoardColor, BoardPieces } from './board';

type ColorShorthand = '⚪' | '⚫';
type PieceTypeShorthand = '🏰' | '🐴' | '⛪' | '👸' | '🤴' | '♟';
type RealPieceShorthand = '⚪🏰' | '⚫🏰' | '⚪🐴' | '⚫🐴' | '⚪⛪' | '⚫⛪' | '⚪👸' | '⚫👸' | '⚪🤴' | '⚫🤴' | '⚪♟' | '⚫♟';
type EmptyCellShorthand = '🕳 🕳';
export type PieceShorthand = EmptyCellShorthand | RealPieceShorthand;

export const toPiece = (pieceShorthand: RealPieceShorthand): Omit<TPiece, 'hasBeenMoved'> => {
    const [colorEmoji, typeEmoji] = pieceShorthand as unknown as [ColorShorthand, PieceTypeShorthand];
    return { color: toPieceColor(colorEmoji), type: toPieceType(typeEmoji) };
}
const toPieceColor = (colorEmoji: ColorShorthand): BoardColor => {
    switch (colorEmoji) {
        case '⚪':
            return 'white';
        case '⚫':
            return 'black';
    }
}
const toPieceType = (pieceTypeEmoji: PieceTypeShorthand): PieceType => {
    switch (pieceTypeEmoji) {
        case '♟':
            return 'pawn';
        case '⛪':
            return 'bishop';
        case '🏰':
            return 'rook';
        case '🐴':
            return 'knight';
        case '👸':
            return 'queen';
        case '🤴':
            return 'king'
    }
}

export const boardToShorthand = (boardPieces: BoardPieces): string => 
    `[\n${boardPieces.map(row => `\t[${row.map(toShorthandFunction).join(', ')}]`).join(',\n')}\n]`

const toShorthandFunction = (piece?: TPiece | null) => piece ? `${piece.hasBeenMoved ? 'm' : 'u'}('${toShorthand(piece)}')` : "e('🕳 🕳')"
export const toShorthand = (piece: TPiece) => `${toShorthandColor(piece.color)}${toShorthandType(piece.type)}` as PieceShorthand

const toShorthandColor = (color: BoardColor): ColorShorthand => {
    switch(color) {
        case 'white':
            return '⚪'
        case 'black':
            return '⚫'
    }
}

const toShorthandType = (type: PieceType): PieceTypeShorthand => {
    switch (type) {
        case 'pawn':
            return '♟';
        case 'bishop':
            return '⛪';
        case 'rook':
            return '🏰';
        case 'knight':
            return '🐴';
        case 'queen':
            return '👸';
        case 'king':
            return '🤴'
    }
}

export const e = (_: EmptyCellShorthand) => undefined;
export const u = (pieceShorthand: RealPieceShorthand): TPiece =>
    ({ ...toPiece(pieceShorthand), hasBeenMoved: false })
export const m = (pieceShorthand: RealPieceShorthand): TPiece =>
    ({ ...toPiece(pieceShorthand), hasBeenMoved: true })