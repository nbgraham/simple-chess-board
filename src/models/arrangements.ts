import { u, e, m } from './piece_shorthand'
import { BoardPieces } from '../utils/board_utils';

export const startingPosition: BoardPieces = [
    [u('⚫🏰'), u('⚫🐴'), u('⚫⛪'), u('⚫👸'), u('⚫🤴'), u('⚫⛪'), u('⚫🐴'), u('⚫🏰')],
    [u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟')],
    [u('⚪🏰'), u('⚪🐴'), u('⚪⛪'), u('⚪👸'), u('⚪🤴'), u('⚪⛪'), u('⚪🐴'), u('⚪🏰')],
]

export const fastestWhiteInCheckMatePosition: BoardPieces = [
    [u('⚫🏰'), u('⚫🐴'), u('⚫⛪'), e('🕳 🕳'), u('⚫🤴'), u('⚫⛪'), u('⚫🐴'), u('⚫🏰')],
    [u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), e('🕳 🕳'), u('⚫♟'), u('⚫♟'), u('⚫♟')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), m('⚫♟'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), m('⚪♟'), m('⚪♟'), m('⚫👸')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), e('🕳 🕳'), e('🕳 🕳'), u('⚪♟')],
    [u('⚪🏰'), u('⚪🐴'), u('⚪⛪'), u('⚪👸'), u('⚪🤴'), u('⚪⛪'), u('⚪🐴'), u('⚪🏰')],
];

export const whitePawnJustReachedEndOfBoardAtRow0Col6: BoardPieces = [
    [u('⚫🏰'), u('⚫🐴'), u('⚫⛪'), u('⚫👸'), u('⚫🤴'), u('⚫⛪'), m('⚪♟'), u('⚫🏰')],
    [u('⚫♟'), u('⚫♟'), u('⚫♟'), u('⚫♟'), e('🕳 🕳'), u('⚫♟'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), m('⚫♟'), m('⚫🐴'), e('🕳 🕳'), m('⚫♟')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), m('⚫♟'), e('🕳 🕳'), e('🕳 🕳')],
    [e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳'), e('🕳 🕳')],
    [u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), u('⚪♟'), e('🕳 🕳'), e('🕳 🕳'), u('⚪♟')],
    [u('⚪🏰'), u('⚪🐴'), u('⚪⛪'), u('⚪👸'), u('⚪🤴'), u('⚪⛪'), u('⚪🐴'), u('⚪🏰')]
]