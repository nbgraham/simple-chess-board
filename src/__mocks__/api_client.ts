import { ChessMove } from '../models/chess_move';
import { RewindableReducerState, SerDeRewindableReducerState, UNDO, RESET, REDO } from '../utils/rewindable_reducer';
import { makeRewindableReducer } from '../utils/rewindable_reducer';
import { makeStartingRewindableState, gameStateReducer } from '../reducers/game_state_reducer';
import { GameStateDto } from '../models/game_state_dto';
import { IApiClient } from '../api_client';

const initialRewindableBoardState = makeStartingRewindableState();
const rewindableReducer = makeRewindableReducer(gameStateReducer, initialRewindableBoardState)

type BoardListener = (s: SerDeRewindableReducerState<GameStateDto, ChessMove>) => void;
type BoardListenersReducerAction = {
    type: 'add',
    listener: BoardListener
} | {
    type: 'remove'
    listener: BoardListener
}
class MockApiClient implements IApiClient {
    gameState: RewindableReducerState<GameStateDto, ChessMove> = initialRewindableBoardState
    boardListeners: Array<BoardListener> = []

    static boardListenerReducer(prevListeners: Array<BoardListener>, action: BoardListenersReducerAction) {
        if (action.type === 'add') {
            return [...prevListeners, action.listener]
        } else if (action.type === 'remove') {
            return prevListeners.filter(l => l !== action.listener)
        } else {
            return prevListeners
        }
    }

    private dispatch(action: BoardListenersReducerAction) {
        this.boardListeners = MockApiClient.boardListenerReducer(this.boardListeners, action)
    }

    private emitBoardUpdate() {
        this.boardListeners.forEach(listener => listener(this.gameState))
    }

    subcribeToBoard(listener: BoardListener) {
        this.dispatch({ type: 'add', listener })
        this.emitBoardUpdate()
        return {
            unsubscribe: () => { this.dispatch({ type: 'remove', listener }) }
        }
    }

    sendMove(move: ChessMove) {
        this.gameState = rewindableReducer(this.gameState, move)
        this.emitBoardUpdate()
    }

    undoMove() {
        this.gameState = rewindableReducer(this.gameState, UNDO.action)
        this.emitBoardUpdate()
    }

    redoMove() {
        this.gameState = rewindableReducer(this.gameState, REDO.action)
        this.emitBoardUpdate()
    }

    resetBoard() {
        this.gameState = rewindableReducer(this.gameState, RESET.action)
        this.emitBoardUpdate()
    }
}

export const API_CLIENT = new MockApiClient()