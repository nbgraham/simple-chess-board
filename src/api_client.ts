import io from 'socket.io-client';
import { ChessMove } from './models/chess_move';
import { RewindableReducerState, SerDeRewindableReducerState } from './utils/rewindable_reducer';
import { SerDeClass, boardDtoAsClass, chessMoveAsClass, gameStateDtoAsClass } from './utils/serde';
import { GameStateDto } from './models/game_state_dto';

const API_URL = 'http://192.168.1.191:8000';

export const chessServerEvents = {
    subscribeToBoard: 'subscribe_to_board',
    updateBoardState: 'update_board_state',
    sendMove: 'send_move',
    undoMove: 'undo_move',
    redoMove: 'redo_move',
    resetBoard: 'reset_board',
}

class ApiClient {
    socket: SocketIOClient.Socket;

    constructor(apiUrl: string) {
        this.socket = io(apiUrl);
    }

    private emit(event: keyof typeof chessServerEvents, ...args: any[]) {
        console.debug('Emit event:', event)
        return this.socket.emit(chessServerEvents[event], ...args)
    }

    private on(event: keyof typeof chessServerEvents, fn: Function) {
        return this.socket.on(chessServerEvents[event], (...args: any[]) => {
            console.debug('Recieved event:', event)
            fn(...args)
        })
    }

    subcribeToBoard(cb: (s: SerDeRewindableReducerState<GameStateDto, ChessMove>) => void) {
        const listener = (boardHistory: RewindableReducerState<SerDeClass<GameStateDto>, ChessMove>) => {
            console.debug('Got new board from server', boardHistory)
            const value = {
                ...boardHistory,
                currentState: gameStateDtoAsClass(boardHistory.currentState),
                pastStates: boardHistory.pastStates.map(boardDtoAsClass),
                futureStates: boardHistory.futureStates.map(boardDtoAsClass),
                pastActions: boardHistory.pastActions.map(chessMoveAsClass),
                futureActions: boardHistory.futureActions.map(chessMoveAsClass)
            }
            cb(value)
        }
        this.on('updateBoardState', listener)
        this.emit('subscribeToBoard')
        return {
            unsubscribe: () => { this.socket.removeListener(chessServerEvents.updateBoardState, listener) }
        }
    }

    sendMove(move: ChessMove) {
        this.emit('sendMove', move)
    }

    undoMove() {
        this.emit('undoMove', '')
    }

    redoMove() {
        this.emit('redoMove', '')
    }

    resetBoard() {
        this.emit('resetBoard', '')
    }
}

export const API_CLIENT = new ApiClient(API_URL)