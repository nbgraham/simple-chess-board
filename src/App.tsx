import React from 'react';
import './App.css';
import { SmartBoard } from './components/board';
import { BoardContextProvider } from './context';
import { ScorePanel } from './components/score_panel';
import { API_CLIENT } from './api_client';

type AppState = {
  hasError: boolean
}
class App extends React.Component<{}, AppState> {
  state = { hasError: false }

  static getDerivedStateFromError(error: Error) {
    console.error(error)
    return {
      hasError: true
    }
  }

  reload = () => this.setState({ hasError: false })

  reset = () => {
    API_CLIENT.resetBoard()
    this.reload();
  }

  render() {
    if (this.state.hasError) {
      return (
        <div style={{ height: '100%', display: 'flex', flexDirection: 'column', justifyContent: 'center' }}>
          <div style={{ height: '20%', display: 'flex', flexDirection: 'column', justifyContent: 'space-around' }}>
            <div style={{ textAlign: 'center' }}>Oops, something went wrong!</div>
            <div style={{ display: 'flex', justifyContent: 'center' }}><button style={{ margin: 'auto', width: '50%', padding: 10 }} onClick={this.reload}>Reload, Keep Game State</button></div>
            <div style={{ display: 'flex', justifyContent: 'center' }}><button style={{ margin: 'auto', width: '50%', padding: 10 }} onClick={this.reset}>Reset, Lose Game State</button ></div>
          </div>
        </div >
      )
    }

    return (
      <BoardContextProvider apiClientType={'server'}>
        <div style={{
          display: 'flex',
          flexDirection: 'row',
          height: '100%',
          width: '100%',
          overflow: 'hidden'
        }}>
          <div style={{ flex: 1, overflow: 'auto' }}>
            <ScorePanel />
          </div>
          <div style={{ flex: 6 }}>
            <SmartBoard />
          </div>
        </div>
      </BoardContextProvider>
    )
  }
}

export default App;
