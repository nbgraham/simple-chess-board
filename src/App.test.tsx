import React from 'react';
import { render, fireEvent, Matcher, act } from '@testing-library/react';
import App from './App';
import { API_CLIENT } from './api_client';
import { u } from './models/piece_shorthand';
import { c } from './models/cell_shorthand';
jest.mock('./api_client')

beforeEach(() => API_CLIENT.resetBoard())

test('renders app', () => {
  const { getByText, getByTestId } = render(<App />);

  expect(getByText(/New Game/i)).toBeInTheDocument();

  expect(getByText(/turn: white/i)).toBeInTheDocument()

  fireEvent.click(getByTestId("Choose White"))
  expectToHavePiece(getByTestId('a1'), 'white rook')
  expectToHaveNoPiece(getByTestId('e5'))
});

test('fastest checkmate from whites perspective', () => {
  const { queryByText, getByText, getByTestId } = render(<App />);
  const movePiece = createMovePiece(getByTestId)

  fireEvent.click(getByTestId("Choose White"))

  expect(queryByText(/turn: white/i)).toBeInTheDocument()
  movePiece('f2', 'white pawn', 'f4')
  expect(queryByText(/undo/i)).toBeInTheDocument()

  expect(queryByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«â™Ÿ'),
      moveFrom: c('e7'),
      moveTo: c('e5')
    })
  })
  expect(queryByText(/undo/i)).not.toBeInTheDocument()

  expect(queryByText(/turn: white/i)).toBeInTheDocument()
  movePiece('g2', 'white pawn', 'g4')
  expect(queryByText(/undo/i)).toBeInTheDocument()

  expect(queryByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«ðŸ‘¸'),
      moveFrom: c('d8'),
      moveTo: c('h4')
    })
  })
  expect(queryByText(/undo/i)).not.toBeInTheDocument()

  expect(getByText(/wins!/i)).toHaveTextContent(/black/i)
});


test('fastest checkmate from black perspective', () => {
  const { queryByText, getByText, getByTestId } = render(<App />);
  const movePiece = createMovePiece(getByTestId)

  fireEvent.click(getByTestId("Choose Black"))

  expect(queryByText(/turn: white/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âšªâ™Ÿ'),
      moveFrom: c('f2'),
      moveTo: c('f4')
    })
  })
  expect(queryByText(/undo/i)).not.toBeInTheDocument()

  expect(queryByText(/turn: black/i)).toBeInTheDocument()
  movePiece('e7', 'black pawn', 'e5')
  expect(queryByText(/undo/i)).toBeInTheDocument()

  expect(queryByText(/turn: white/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âšªâ™Ÿ'),
      moveFrom: c('g2'),
      moveTo: c('g4')
    })
  })
  expect(queryByText(/undo/i)).not.toBeInTheDocument()

  expect(queryByText(/turn: black/i)).toBeInTheDocument()
  movePiece('d8', 'black queen', 'h4')
  expect(queryByText(/undo/i)).toBeInTheDocument()

  expect(getByText(/wins!/i)).toHaveTextContent(/black/i)

  expectToHavePiece(getByTestId('h4'), 'black queen')
  expectToHaveNoPiece(getByTestId('d8'))

  fireEvent.click(getByText(/undo/i))

  expectToHaveNoPiece(getByTestId('h4'))
  expectToHavePiece(getByTestId('d8'), 'black queen')

  expect(queryByText(/turn: black/i)).toBeInTheDocument()
});


test('resume from API', () => {
  const { getByText, getByTestId, unmount } = render(<App />);
  fireEvent.click(getByTestId("Choose White"))

  const movePiece = createMovePiece(getByTestId)

  expect(getByText(/turn: white/i)).toBeInTheDocument()
  movePiece('f2', 'white pawn', 'f4')

  expect(getByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«â™Ÿ'),
      moveFrom: c('e7'),
      moveTo: c('e5')
    })
  })

  unmount()

  const { getByTestId: getByTestId1, unmount: unmount1 } = render(<App />);
  fireEvent.click(getByTestId("Choose White"))

  expectToHavePiece(getByTestId1('f4'), 'white pawn')
  expectToHavePiece(getByTestId1('e5'), 'black pawn')
  unmount1()

  act(() => {
    API_CLIENT.resetBoard()
  })

  const { getByTestId: getByTestId2 } = render(<App />);
  fireEvent.click(getByTestId("Choose White"))

  expectToHaveNoPiece(getByTestId2('f4'))
  expectToHaveNoPiece(getByTestId2('e5'))
});

test('redo', () => {
  const { queryByText, getByText, getByTestId, asFragment } = render(<App />);
  fireEvent.click(getByTestId("Choose Black"))

  const movePiece = createMovePiece(getByTestId)

  expect(getByText(/turn: white/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âšªâ™Ÿ'),
      moveFrom: c('f2'),
      moveTo: c('f4')
    })
  })

  expect(getByText(/turn: black/i)).toBeInTheDocument()
  movePiece('e7', 'black pawn', 'e5')

  expect(queryByText(/redo/i)).not.toBeInTheDocument()

  fireEvent.click(getByText(/undo/i))
  expectToHaveNoPiece(getByTestId('e5'))

  fireEvent.click(getByText(/redo/i))
  expectToHavePiece(getByTestId('e5'), 'black pawn')
})

test('undo after resume', () => {
  const { getByText, getByTestId, unmount } = render(<App />);
  const movePiece = createMovePiece(getByTestId)

  fireEvent.click(getByTestId("Choose White"))

  expect(getByText(/turn: white/i)).toBeInTheDocument()
  movePiece('f2', 'white pawn', 'f4')

  expect(getByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«â™Ÿ'),
      moveFrom: c('e7'),
      moveTo: c('e5')
    })
  })
  
  unmount()

  const { getByText: getByText1, getByTestId: getByTestId1 } = render(<App />);
  fireEvent.click(getByTestId("Choose Black"))

  expectToHavePiece(getByTestId1('f4'), 'white pawn')
  expectToHavePiece(getByTestId1('e5'), 'black pawn')

  fireEvent.click(getByText1(/undo/i))

  expectToHavePiece(getByTestId1('f4'), 'white pawn')
  expectToHaveNoPiece(getByTestId1('e5'))
})

test('new game resets', () => {
  const { getByText, getByTestId, asFragment } = render(<App />);
  const initialRender = asFragment()

  const movePiece = createMovePiece(getByTestId)

  fireEvent.click(getByTestId("Choose White"))

  expect(getByText(/turn: white/i)).toBeInTheDocument()
  movePiece('f2', 'white pawn', 'f4')

  expect(getByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«â™Ÿ'),
      moveFrom: c('e7'),
      moveTo: c('e5')
    })
  })

  expectToHavePiece(getByTestId('f4'), 'white pawn')
  expectToHavePiece(getByTestId('e5'), 'black pawn')

  fireEvent.click(getByText(/new game/i))

  expect(initialRender).toMatchSnapshot(asFragment())
  expectToHaveNoPiece(getByTestId('f4'))
  expectToHaveNoPiece(getByTestId('e5'))
})

test('new game resets when initial mount resumed game', () => {
  const { getByText, getByTestId, unmount } = render(<App />);
  const movePiece = createMovePiece(getByTestId)

  fireEvent.click(getByTestId("Choose White"))

  expect(getByText(/turn: white/i)).toBeInTheDocument()
  movePiece('f2', 'white pawn', 'f4')

  expect(getByText(/turn: black/i)).toBeInTheDocument()
  act(() => {
    API_CLIENT.sendMove({
      type: 'move',
      piece: u('âš«â™Ÿ'),
      moveFrom: c('e7'),
      moveTo: c('e5')
    })
  })

  expectToHavePiece(getByTestId('f4'), 'white pawn')
  expectToHavePiece(getByTestId('e5'), 'black pawn')

  unmount()

  const { getByText: getByText1, getByTestId: getByTestId1 } = render(<App />);

  fireEvent.click(getByTestId("Choose White"))

  expectToHavePiece(getByTestId1('f4'), 'white pawn')
  expectToHavePiece(getByTestId1('e5'), 'black pawn')

  fireEvent.click(getByText1(/new game/i))

  expectToHaveNoPiece(getByTestId1('f4'))
  expectToHaveNoPiece(getByTestId1('e5'))
})

const createMovePiece = (getByTestId: (text: Matcher) => HTMLElement) => (from: string, piece: string, to: string) => {
  expectToHavePiece(getByTestId(from), piece)

  fireEvent.click(getByTestId(from))
  fireEvent.click(getByTestId(to))

  expectToHaveNoPiece(getByTestId(from))
  expectToHavePiece(getByTestId(to), piece)
}

const expectToHaveNoPiece = (cell: HTMLElement) => expect(cell.childElementCount).toBe(0)
const expectToHavePiece = (cell: HTMLElement, piece: string) =>
  expect(cell.firstChild).toHaveAttribute('alt', piece)