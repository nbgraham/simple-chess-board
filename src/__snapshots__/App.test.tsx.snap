// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new game resets 1`] = `Object {}`;

exports[`redo 1`] = `Object {}`;
